src/main/java/io/cdap/plugin/ariba/source/AribaServices.java

```diff
@@ -0,0 +1,383 @@
+import java.util.Base64;
+import java.util.List;
+import java.util.Objects;
+import java.util.concurrent.TimeUnit;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import javax.ws.rs.core.MediaType;
+
+/**
+ * This {@code AribaServices} contains all the Ariba relevant service call implementations
+ * - check the correctness of the formed URL
+ * - builds the Output Schema
+ */
+public class AribaServices {
+
+  public static final String DOCUMENT_TYPE = "documentType";
+  public static final String AUTHORIZATION = "Authorization";
+  public static final String API_KEY = "apiKey";
+  public static final String ACCEPT = "Accept";
+  public static final String POST = "POST";
+  public static final String CONTENT_TYPE = "Content-Type";
+  public static final String VIEW_TEMPLATE = "viewTemplates";
+  public static final String PRODUCT = "product";
+  public static final String REALM = "realm";
+  public static final String URL_PATTERN = "(api\\S*?).com";
+  public static final String TOKEN_PATH_SEGMENT = "v2/oauth/token";
+  public static final String TOKEN_GRANT_TYPE = "grant_type=client_credentials";
+  public static final String ANALYTICS_PATH_SEGMENT = "api/analytics-reporting-view/v1";
+  public static final String ANALYTICS = "analytics";
+  public static final String METADATA = "metadata";
+  public static final String ACCESS_TOKEN = "access_token";
+
+  private static final Logger LOG = LoggerFactory.getLogger(AribaServices.class);
+  private final AribaPluginConfig pluginConfig;
+
+  public AribaServices(AribaPluginConfig pluginConfig) {
+    this.pluginConfig = pluginConfig;
+  }
+
+  /**
+   * Calls to check Ariba Connection.
+   */
+  public String getAccessToken() throws AribaException, IOException {
+    LOG.trace("Initiating Ariba connection for access token");
+    AribaResponseContainer responseContainer = callAribaForToken(generateTokenURL());
+    InputStream responseStream = responseContainer.getResponseStream();
+    ObjectMapper objectMapper = new ObjectMapper();
+    if (responseContainer.getHttpStatusCode() == HttpURLConnection.HTTP_OK) {
+      return objectMapper.readTree(responseStream).get(ACCESS_TOKEN).asText();
+    }
+    String errMsg = ResourceConstants.ERR_FETCHING_TOKEN.getMsgForKey();
+    throw new AribaException(errMsg, responseContainer.getHttpStatusCode());
+  }
+
+  /**
+   * Construct metadata URL.
+   *
+   * @return URL.
+   */
+  @VisibleForTesting
+  protected URL generateTokenURL() {
+    String tokenUrl = fetchAuthURL();
+    String newTokenUrl = (tokenUrl.startsWith("https://")) ? tokenUrl : "https://" + tokenUrl;
+    HttpUrl.Builder builder = Objects.requireNonNull(HttpUrl.parse(newTokenUrl))
+      .newBuilder().addPathSegments(TOKEN_PATH_SEGMENT);
+    return builder.build().url();
+  }
+
+  /**
+   * Fetches Authentication url from base url
+   *
+   * @return url string
+   */
+  private String fetchAuthURL() {
+    Pattern urlPattern = Pattern.compile(URL_PATTERN);
```

Comment:
This is the actual endpoint: https://openapi.au.cloud.ariba.com
And only for Token call we need to fetch this(https://api.au.cloud.ariba.com) url from above url, rest all the call will be happening on base url only.

Changed the default behaviour to AribaException if regex doesn't match.

Reply:
Answer: Yes, it true for all the customers and is mentioned on the developer portal itself. Pasting the link but it might require credential to log in to the below link.
Link: https://developer.au.cloud.ariba.com/api/apis

Reply:
Sent you mail with Screenshots and document as this URL requires credentials.
Mail Subject "OAuth related ask from PR".

Reply:
Base URL: https://openapi.au.cloud.ariba.com
Auth URL: https://api.au.cloud.ariba.com